User accounts

opadron opened the pull request:
Adds frontend UI components and logic for registering new accounts and logging into registered accounts.

Also significantly reorganizes and simplifies the frontend code base.  See commit messages for more details.

Commit:
significantly simplify application state tree/reducer

 - rework compose functionality to reduce the number of closures needed.
 - update compose test
 - add common reducers to common-reducers module

Commit:
refactor components into presentational and container components

 - add new common components: group and text-field
 - rework global-nav components so that instead of using an explicit table of
 components, the table is implied by the component's children, which are
 highlighted using the current navTarget as the key.
 - remove prop-types module; instead, use the React PropTypes object directly.
 - remove unused front-page components

Commit:
fix Promise-wrapped ajax function

 - fixed so that a failure in the underlying jQuery call results in the promise
 being rejected with a proper Error value.

Commit:
refactor/add dependencies

 - add new dependency: redux-logger
 - move app-wide operations to their own "actions" module
 - move app-wide data structures to their own "globals" module
 - move app-wide route handlers to their own "routes" module

opadron commented:
@osumo/developers any ideas?  Looking around, I've found nodejs/node-v0.x-archive#5504, but not much else.

jeffbaumes commented:
To me it looks like it may have something to do with [routers](https://github.com/ReactTraining/react-router/pull/222) or other object which have many [listeners](https://pouchdb.com/errors.html#event_emitter_limit). It could also be a true memory leak. Could your router object have many (more than 10) listeners attached to it? The first router link is a PR that makes the limit infinite. Seems a bit wrong to do that - perhaps we should see if increasing to 20 helps?

opadron commented:
@osumo/developers can someone QA this?

waxlamp commented on web-external/src/components/dialog/login/footer.jsx:
```diff
@@ -0,0 +1,37 @@
+                type='submit'
+                onClick={ onSubmit }>
+          <i className='icon-login'/>
+          { NBSP } Login
```
s/Login/Log in/

waxlamp replied:
Get rid of the `NBSP` (looks weird on the button).

waxlamp commented on web-external/src/components/dialog/register/footer.jsx:
```diff
@@ -0,0 +1,35 @@
+                id='g-register-button'
+                type='submit'
+                onClick={ onSubmit }>
+          <i className='icon-plus-circled'>Register</i>
```
This makes `Register` appear in italics. Follow the example of `login/footer.jsx` and put a slash before the `i` tag's closing bracket, and remove the `</i>`.

waxlamp reviewed (changes requested):
components/header/user.jsx, lines 45, 51, and 58 have the same "italic icon" bug as documented in the line notes for the register dialog.

Commit:
fix register/log in buttons (@ronichoudhury)

Commit:
Merge branch 'master' into user-accounts

Commit:
try to fix functionality issues introduced in the merge

There's an issue where every time a file in the new
file-selection widget is clicked, the user is
redirected to the main page instead of having the file
added to the select element.

opadron commented:
I'm trying to merge this with the latest in master after landing #30.  It looks like I'm running into an issue I don't fully understand: it seems like every time I click on a file in the new file-selection widget, I end up getting redirected to the main page.

Commit:
refactor and add backupKey feature to common Group component

Commit:
complete integration of file selector widget

 - better encapsulated the girder HeirarchyWidget view.

   - Now, the Backbone view should be fully wrapped by the
     file-selector body component, and the header and other
     subcomponents have been slightly restructured to match
     the organization of the other dialog components.

 - moved item selected callback logic out of the new Select
   component; instead, there is now only one file selection
   dialog component (that is integrated with the other
   dialog components) and one global item selected callback
   that uses one additionional level of indirection so that
   other components can customize what happens when a user
   clicks on a file.  Now, the Select component is much
   simpler; just a few divs, an input and a button.

   - added new global variable: itemSelectedCallback

   - added new global action: onItemSelect

   - added new global action: openFileSelectorDialog

   - added new global action: setItemSelectedCallback

     - actions.onItemSelect() always forwards to
       globals.itemSelectedCallback().  To use the new
       file-selection component, a custom callback that
       dictates what to do when an item is clicked should be
       provided up-front using
       actions.setItemSelectedCallback(), then the
       file-selection dialog can be opened by calling
       actions.openFileSelectorDialog().  When the user
       clicks on an item the provided callback is called,
       which should close the dialog and proceed to process
       the chosen item.

   - added new route: 'file-select'

     - NOTE: since we should really programmatically set the
       item selected callback beforehand, it probably
       doesn't make sense for this to have a route, unless
       we work out a default handler.

 - removed g-dialog-container div in index.html

 - removed pug templates

 - moved select component code so that it lives under
   src/components/common.

 - added new code to entry point

   - girder.router.navigate = () => null;

   - This is an attempt to stub out whatever it is the
     girder library is doing to meddle with the location
     hash.  Since we have our own routing solution with our
     own router, we should handle routing through that.

     - NOTE: actually, come to think of it, it would
       probably work better to have the girder router
       defer to ours (girder.router.navigate = ourRouter).
       Some experimentation would be good here.

 - added new global action: setFileNavigation

 - added new redux state variable: dialog.currentFolder

 - added new redux state variable: dialog.currentFolder

   - girder user, girder collection, or girder folder
     object

 - added new redux state variable:
   dialog.currentFolderParentType

   - string set to one of 'user', 'collection', or 'folder'

   - together, these state variables control where the
     file-select navigation dialog starts every time it is
     opened.  NOTE: we should think about how to "remember"
     where the user left off from the last time they used
     the dialog.  Use actions.setFileNavigation(type, id) to
     update this part of the state.

Commit:
fix Group component and include onlyMatching option

Commit:
fix an issue causing bad requests in file-selector component

opadron commented:
I've managed to figure out a lot of the kinks and get this work well integrated with #30.  I think it's working really well!  The only quirk I still notice is that the page scrolls to the bottom whenever you finish using the file-selector component.

@ronichoudhury please take a look at these latest commits, 9d4f9fd in particular.  I've summarized the majority of my latest contributions there.

ronichoudhury commented:
I am still seeing italic text alongside icons, etc. (look, for instance, inside the Girder "username" dropdown menu at the right of the navbar).

ronichoudhury commented:
I am having some trouble understanding the latest changes - but it's mainly a matter of my lack of background/depth with React.

My intuition is still telling me that something as simple as adding a file selector widget shouldn't cause such a dispersed changeset. I will take a closer look again in a bit.

ronichoudhury commented:
Another comment - I noticed that the file selector doesn't *quite* work the way it was before. Now, when I activate it, it starts off in my home folder, and I have no way to get to the collection containing the OSUMO input stuff. Is this by design?

opadron commented:
> I am still seeing italic text alongside icons, etc. (look, for instance, inside the Girder "username" dropdown menu at the right of the navbar).

Sorry, I did all this integration work and completely forgot to address your original points.  I'll add the fixes, soon.

> My intuition is still telling me that something as simple as adding a file selector widget shouldn't cause such a dispersed changeset.

I think we should be careful to distinguish between dispersion over actual files and lines of code, and conceptual dispersion (both being present in these changesets).  The first we could always change by reorganizing how the code is laid out, but I would argue the latter is a direct consequence of the core tenets of React + Redux development, and we won't be able to do anything about it short of switching to something else.

Even for the simplest widgets, you're more likely than not to have to write code for the widget(s) themselves; code for container widget(s) to tie their props to the redux state tree; code that adds the new parts of the state tree that will be used in evaluating said props, including their reducers; code that implements the new actions that the widget(s) will trigger in their callbacks, and any new routes that would be mapped to these new actions.  We could move code around to keep changes limited to fewer files and lines of code at a time, but conceptually, there's a much more fundamental dispersion in React + Redux development that can't be helped.  Personally, I found that once I took the time to understand these patterns, it was much easier for me to go through these seemingly dispersed change sets and quickly make sense of them.

It is my sincere hope that you and @manthey would come to agree, but if you think the development style is really going to continue making things difficult for you, we should take a step back and reevaluate, while we still can.  Better for us to share an MO that we can all work with.

Commit:
fix more italics mistakes

opadron commented:
@ronichoudhury hopefully, I got all my italics slip ups.

Regarding the file browsing, I did tweak the behavior so that it always starts in your home space.  For my local instance, the data is under an `osumopublicuser` user's space, but if you have your data in a collection, then you are absolutely right that navigating to that is not possible in the UI (I think).

The machinery is in place to handle that navigation programmatically:

```javascript
actions.setFileNavigation('collection', collectionId);
```

but as far as I can tell, the HierarchyView doesn't have the UI components needed to browse from collection to collection, or from one user's space to another.  It would be nice if there were a way to navigate to some "root" view, where users can see all the collections and users that they have permission to view.

I'd love to know what your thoughts are.

ronichoudhury commented:
I uploaded a file to my user space so I could test, and everything seems to be working.

To make this approach general, we might want to have a script similar to https://github.com/Kitware/candela/blob/master/app/resonantlab/scripts/populateGirder.py to upload some data into some account's home folder.

I am not sure why the page jumps like that after coming back from the file selector, but the problem is not unique to that modal. If you do something like bring up the login dialog and then dismiss it, you will get the same page jumping effect. i.e., this seems to be connected to Girder modals. It doesn't happen on data.kitware.com, though, so I'm not sure what that means for OSUMO.

manthey commented:
Should we be using the Browser widget rather than the Hierarchy widget?  This would let someone switch roots (different collections or users).

ronichoudhury commented:
That sounds a lot like what we should be using :). Can that be "modalized"
the same way we have done with HierarchyWidget?

On Wed, Nov 9, 2016 at 3:39 PM David Manthey <notifications@github.com>
wrote:

> Should we be using the Browser widget rather than the Hierarchy widget?
> This would let someone switch roots (different collections or users).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/osumo/osumo/pull/24#issuecomment-259520275>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACxNJGaRg0wybL3T7djIKlOCQJtkrARSks5q8i-OgaJpZM4J8AX4>
> .
>

opadron commented:
See [this](https://gist.github.com/opadron/8cd3f190eb2c77bf2017e9e1790ba363) patch (based on c2eb4c2) for my attempt at using multiple root divs to address the scrolling issue.  It looks like hoisting the dialog components to their own React component tree doesn't change the scrolling behavior.

I've checked all the computed values for the styling for pre-patch, post-patch, and compared it all to Girder's own styling and everything matches across all three cases, so I don't think it's a styling issue.

Commit:
hoist dialog components/wrap root selector component

opadron commented:
@manthey Check at the latest (c652548) changes.  I tried different approaches to using the BrowserWidget.  The solution I ended up going with was to just add the root selector piece.

Besides working pretty well, I'm also learning a lot about working Backbone into a React (+ Redux) app, which should help clear the way for more integration with girder core components.

Still not sure what to do about the scroll-jumping quirk.

ronichoudhury commented:
We'll have to look a bit deeper into that "jumping" issue. I suspect it still has to do with the nesting issue @manthey identified. Maybe we can use a `position` style directive on the overlay CSS to influence how its drawn, etc., even when it's nested as it is with a React component. @manthey may have better ideas than me about that.
